File: Data Structures & Algorithms/anagram-groups/submission-5.py
class Solution:
    def is_Anagrams(self, s:str, t:str) -> bool:
        if len(s) != len(t):
            return False
        mp1 = {}
        mp2 = {}
        for i in range(len(s)):
            mp1[s[i]] = mp1.get(s[i],0) + 1
            mp2[t[i]] = mp2.get(t[i],0) + 1
        return mp1 == mp2

    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        if len(strs) == 0:
            return []
        if len(strs) == 1:
            return [strs]
        ans=[]
        while len(strs) > 0:
            res = []
            pivot = strs[0]
            res.append(pivot)
            del strs[0]
            j = 0
            while j < len(strs):
                word = strs[j]
                if self.is_Anagrams(pivot, word):
                    res.append(word)
                    strs.pop(j)
                else:
                    j += 1
            ans.append(res)
        return ans

File: Data Structures & Algorithms/count-subsequences/submission-5.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int n = s.size();
        int m = t.size();
        vector<vector<int>> dp(n + 1,vector<int>(m + 1,0));
        for(int i = 0;i <= n;i++){
            for(int j = 0;j <= m;j++){
                if (j == 0){
                    dp[i][j] = 1;
                    continue;
                }
                if (i == 0){
                    dp[i][j] = 0;
                    continue;
                }
                dp[i][j] = dp[i - 1][j];
                if (s[i - 1] == t[j - 1]){
                    dp[i][j] += dp[i - 1][j - 1];
                }
            }
        }
        return dp[n][m];
    }
};
